src/en/dynasty/src/eu/kanade/tachiyomi/extension/en/dynasty/Dynasty.kt

```diff
@@ -0,0 +1,698 @@
+            if (it == "all") {
+                Int.MAX_VALUE
+            } else {
+                it.toInt()
+            }
+        }
+
+    private val covers: Map<String, Map<String, String>> by lazy {
+        this::class.java
+            .getResourceAsStream("/assets/covers.json")!!
+            .bufferedReader().use { it.readText() }
+            .parseAs()
+    }
+
+    private fun getCoverUrl(directory: String?, permalink: String): String? {
+        directory ?: return null
+
+        if (directory == "chapters") {
+            return buildChapterCoverFetchUrl(permalink)
+        }
+
+        val file = covers[directory]?.get(permalink)
+            ?: return null
+
+        return buildCoverUrl(file)
+    }
+
+    private fun buildCoverUrl(file: String): String {
+        return baseUrl.toHttpUrl()
+            .newBuilder()
+            .addEncodedPathSegments(
+                file.removePrefix("/")
+                    .substringBefore("?"),
+            )
+            .fragment(COVER_URL_FRAGMENT)
+            .build()
+            .toString()
+    }
+
+    private fun buildChapterCoverFetchUrl(permalink: String): String {
+        return HttpUrl.Builder().apply {
+            scheme("http")
+            host(COVER_FETCH_HOST)
+            addQueryParameter("permalink", permalink)
+        }.build().toString()
+    }
+
+    private fun fetchCoverUrlInterceptor(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+
+        if (request.url.host != COVER_FETCH_HOST) {
+            return chain.proceed(request)
+        }
+
+        val permalink = request.url.queryParameter("permalink")!!
+
+        val chapterUrl = baseUrl.toHttpUrl().newBuilder().apply {
+            addPathSegment("chapters")
+            addPathSegments("$permalink.json")
+        }.build()
+
+        val page = client.newCall(GET(chapterUrl, headers)).execute()
+            .parseAs<ChapterResponse>()
+            .pages.first()
+
+        val url = buildCoverUrl(page.url)
+
+        val newRequest = request.newBuilder()
+            .url(url)
+            .build()
+
+        return chain.proceed(newRequest)
+    }
+
+    private fun coverInterceptor(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+
+        return if (request.url.fragment == COVER_URL_FRAGMENT) {
+            coverClient.newCall(request).execute()
+        } else {
+            chain.proceed(request)
+        }
+    }
+
+    private fun String.permalinkToTitle(): String {
+        val result = StringBuilder(length)
+        var capitalize = true
+        for (char in this) {
+            result.append(
+                if (char == '_') {
+                    ' '
+                } else if (capitalize) {
+                    char.uppercase()
+                } else {
+                    char.lowercase()
+                },
+            )
+
+            capitalize = char == '_'
+        }
+
+        return result.toString()
+    }
```

Comment:
Seems overly complicated. I assume what you want is `something_like_this` to become `Something Like This`?
```kt
split('_')
    .joinToString(" ") { word ->
        word.replaceFirstChar { it.uppercase() }
    }
```
(or `word.lowercase().replaceFirstChar...`)

Reply:
mostly for micro-optimization of only looping once, but will change

Reply:
Here I would say the tradeoff with maintainability is not worth it.